### ISSUE TITLE
False-positive `dead_code` warning for code only used in compile-time assertions

### ISSUE THREAD

**Issue by pr2502** [rust-lang/rust#89717]
Sometimes when I want to assert some safety invariant I'll write a constant that asserts it at compile time instead of a test to ensure my code won't compile if it's violating that invariant.

Given the following code: <https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5c02e5e11f9bd262648dd7fb92dd9b81>

```rust
const fn one() -> usize {
    1
}

const _CT_ASSERT: () = {
    assert!(one() == 2);
};
```

The current output is:

```
warning: function is never used: `one`
 --> src/lib.rs:1:10
  |
1 | const fn one() -> usize {
  |          ^^^
  |
  = note: `#[warn(dead_code)]` on by default

error[E0080]: evaluation of constant value failed
 --> src/lib.rs:6:5
  |
6 |     assert!(one() == 2);
  |     ^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: one() == 2', src/lib.rs:6:5
  |
  = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0080`.
```

I'd expect rustc to not emit the `dead_code` warning because the function is clearly being used at compile time, the assert even fails in this case.

It is possible that I'm misusing const eval here and that the code in the constant initializer is not guaranteed to run at compile time if the constant itself is not used and therefore all code used in its initializer is not considered used either. (The warning does disappear when i use the constant itself, which would be consistent with this.) In which case I'd want to turn this into a question, is there any other way to write compile-time assertions like this one?

**Comment by IsaiahCoroama** [comment]
Can be reproduced without an assertion as well:
### Repro
```rust
const fn one() -> usize {
    1
}

const _CT_ASSERT: () = {
    let _ = one();
};
```
### Current Output
```rust
warning: function `one` is never used
 --> src/lib.rs:1:10
  |
1 | const fn one() -> usize {
  |          ^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default

warning: 1 warning emitted
```
Changing the name from `const _CT_ASSERT` to `const _` compiles with no warnings.

**Comment by IsaiahCoroama** [comment]
Given the `const _` case is handled for this, should it extend to named consts too?
